### PR TITLE
Initial session pool

### PR DESCRIPTION
**Pull request by chbatey**
Needs more work but be good to get this in so we can work in parallel on everything else

Refs #4

### CODE REVIEW

File: journal/src/main/resources/reference.conf [hunk 1]
```diff
@@ -4,6 +4,22 @@ akka.persistence.spanner {
  project = "akka"
  instance = "akka"
  database = "akka"
+
+
+ session-pool {
+  # the max number of sessions to create when connecting to spanner.
+  # currently all of these sessions are created eagerly and kept alive.
+  # called 'max-sessions' as future versions will likely a dynamic
```
Comment (johanandren) [inline journal/src/main/resources/reference.conf:12]: lol, no it is not called `max-sessions`

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        unusedSessions = session :: unusedSessions
+        stash.unstash(this, 1, identity)
+      } else {
+        log.error(
+          "unknown session returned {}. This is a bug. Used sessions {}. Unused sessions {}",
+          id,
+          usedSessions,
+          unusedSessions
+        )
+        Behaviors.same
```
Comment (johanandren) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:144]: this

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+      }
+    case other =>
+      log.error("Unexpected message in active state {}", other)
+      Behaviors.same
```
Comment (johanandren) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:148]: this

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+@InternalApi
+private[spanner] final class SessionPool(
+    initialSessions: List[Session],
+    ctx: ActorContext[Command],
+    timers: TimerScheduler[Command],
+    stash: StashBuffer[Command]
+) extends AbstractBehavior[SessionPool.Command](ctx) {
+  private val log = ctx.log
+  private var unusedSessions = initialSessions
+  private var usedSessions = Map.empty[UUID, Session]
```
Comment (johanandren) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:112]: Used/unused feels a bit off, what about `availableSessions` or `freeSessions` and `inUseSessions` instead?

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+  override def onMessage(msg: Command): Behavior[Command] = msg match {
+    case gt: GetSession =>
```
Comment (johanandren) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:115]: match using unapply would be cleaner?

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+ *  - deal with broken session
+ *  - exponential back off for session creation
+ *
+ * See google advice for session pools: https://cloud.google.com/spanner/docs/sessions#create_and_size_the_session_cache
+ *
+ */
+@InternalApi
+private[spanner] object SessionPool {
+  sealed trait Command
+  final case class GetSession(replyTo: ActorRef[Response], id: UUID) extends Command
```
Comment (johanandren) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:37]: So the idea is that the client of the pool creates an UUID to identify the request-response. Could it be more efficient if the pool assigned/created an id for each session up front and returned session+id instead (and re-used it for the lifetime of the session)?

File: journal/src/test/scala/akka/persistence/spanner/internal/AbstractStubbedSpannerClient.scala [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  PartitionQueryRequest,
+  PartitionReadRequest,
+  PartitionResponse,
+  ReadRequest,
+  ResultSet,
+  RollbackRequest,
+  Session,
+  SpannerClient,
+  Transaction
+}
```
Comment (johanandren) [inline journal/src/test/scala/akka/persistence/spanner/internal/AbstractStubbedSpannerClient.scala:34]: Epic import (would perhaps justify a wildcard import?)

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala (hunk 6 above)
Comment (chbatey) [reply journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:37]: The reason i did it this way was if a request to the pool times out a client can still cancel its request without ever receiving a response

Comment (johanandren) [reply journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:37]: Ah. I see. Let's keep as is and see how it works out when we implement the rest.

File: journal/src/main/resources/reference.conf (hunk 1 above)
Comment (chbatey) [reply journal/src/main/resources/reference.conf:12]: Doh, that had a rename

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala (hunk 4 above)
Comment (chbatey) [reply journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:112]: 👍

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala (hunk 6 above)
Comment (chbatey) [reply journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:37]: An alternative would be that we did it the way you suggest and have a timeout in the pool for unreturned sessions

File: journal/src/test/scala/akka/persistence/spanner/internal/AbstractStubbedSpannerClient.scala (hunk 7 above)
Comment (chbatey) [reply journal/src/test/scala/akka/persistence/spanner/internal/AbstractStubbedSpannerClient.scala:34]: I take no responsibility for writnggg this

File: journal/src/main/resources/reference.conf [hunk 8]
```diff
@@ -4,6 +4,22 @@ akka.persistence.spanner {
  project = "akka"
  instance = "akka"
  database = "akka"
+
+
+ session-pool {
+  # the max number of sessions to create when connecting to spanner.
+  # currently all of these sessions are created eagerly and kept alive.
+  # called 'max-size' as future versions will likely a dynamic
```
Comment (patriknw) [inline journal/src/main/resources/reference.conf:12]: will likely have a

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  private final case class RetrySessionCreation(in: FiniteDuration) extends Command
+
+  sealed trait Response {
+    val id: UUID
+  }
+  final case class PooledSession(session: Session, id: UUID) extends Response
+  final case class PoolBusy(id: UUID) extends Response
+
+  def apply(client: SpannerClient, settings: SpannerSettings): Behavior[SessionPool.Command] =
+    Behaviors.withStash[Command](settings.sessionPool.maxOutstandingRequests) { stash =>
```
Comment (patriknw) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:50]: good to use Typed in this new stuff

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+      Behaviors.withTimers[Command] { timers: TimerScheduler[Command] =>
+        Behaviors.setup[Command] { ctx =>
+          ctx.log.info(
+            "Creating pool. Max size {}. Stash {}.",
```
Comment (patriknw) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:54]: use our normal convention of `[{}]`?

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+          Behaviors.receiveMessagePartial {
+            case InitialSessions(sessions) =>
+              stash.unstashAll(new SessionPool(sessions, ctx, timers, stash))
+            case RetrySessionCreation(when) =>
+              if (when == Duration.Zero) {
+                ctx.log.info("Retrying session creation")
+                createSessions()
+              } else {
+                ctx.scheduleOnce(when, ctx.self, RetrySessionCreation(Duration.Zero))
```
Comment (patriknw) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:84]: reason for not using timers?

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+          id,
+          inUseSessions,
+          availableSessions
+        )
+        this
+      }
+    case other =>
+      log.error("Unexpected message in active state {}", other)
+      this
+  }
```
Comment (patriknw) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:149]: need any closing in PostStop (and PreRestart)?

File: journal/src/test/scala/akka/persistence/spanner/internal/SessionPoolStubbedSpec.scala [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+    override def batchCreateSessions(in: BatchCreateSessionsRequest): Future[BatchCreateSessionsResponse] = {
+      val promise = Promise[BatchCreateSessionsResponse]
+      probe.ref ! BatchSessionCreateInvocation(in, promise)
+      promise.future
+    }
+  }
+}
+
+class SessionPoolStubbedSpec extends AnyWordSpecLike with BeforeAndAfterAll with Matchers {
+  val testkit = ActorTestKit()
```
Comment (patriknw) [inline journal/src/test/scala/akka/persistence/spanner/internal/SessionPoolStubbedSpec.scala:39]: use ScalaTestWithActorTestKit instead? (then no need for afterAll)

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+  sealed trait Response {
+    val id: UUID
+  }
+  final case class PooledSession(session: Session, id: UUID) extends Response
+  final case class PoolBusy(id: UUID) extends Response
+
+  def apply(client: SpannerClient, settings: SpannerSettings): Behavior[SessionPool.Command] =
+    Behaviors.withStash[Command](settings.sessionPool.maxOutstandingRequests) { stash =>
+      Behaviors.withTimers[Command] { timers: TimerScheduler[Command] =>
```
Comment (hseeberger) [inline journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:51]: Is the type annotation for `timers` really needed here?

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala (hunk 11 above)
Comment (chbatey) [reply journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:84]: Nope, will switch

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala (hunk 12 above)
Comment (chbatey) [reply journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:149]: Yes. I have created https://github.com/akka/akka-persistence-spanner/issues/12 to follow up

File: journal/src/test/scala/akka/persistence/spanner/internal/SessionPoolStubbedSpec.scala (hunk 13 above)
Comment (chbatey) [reply journal/src/test/scala/akka/persistence/spanner/internal/SessionPoolStubbedSpec.scala:39]: Switched that in my other branch

File: journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala (hunk 14 above)
Comment (chbatey) [reply journal/src/main/scala/akka/persistence/spanner/internal/SessionPool.scala:51]: Nope

**Comment by chbatey** [comment]
A fewer of the smaller things I'll follow up in my next PR

**Pull request merged**